From ffde09e127c3592ad3b1feefc2530901cefe8471 Mon Sep 17 00:00:00 2001
From: Josua Mayer <josua@solid-run.com>
Date: Tue, 25 Oct 2022 12:42:35 +0300
Subject: [PATCH] thermal: nxp_tmu: fix read_remperature function

Backport behaviour from NXP U-Boot lf-5.15.32-2.0.0:
- Use correct check if reported temperature is valid
- abort after 10 retries to avoid infinite loop under error conditions

Signed-off-by: Josua Mayer <josua@solid-run.com>
---
 drivers/thermal/nxp_tmu.c | 24 ++++++++++++++++++++----
 1 file changed, 20 insertions(+), 4 deletions(-)

diff --git a/drivers/thermal/nxp_tmu.c b/drivers/thermal/nxp_tmu.c
index ae0cb65ff5..bced34fb43 100644
--- a/drivers/thermal/nxp_tmu.c
+++ b/drivers/thermal/nxp_tmu.c
@@ -106,15 +106,31 @@ static int read_temperature(struct udevice *dev, int *temp)
 	struct nxp_tmu_plat *pdata = dev_get_platdata(dev);
 	ulong drv_data = dev_get_driver_data(dev);
 	u32 val;
+	u32 retry = 10;
+	u32 valid = 0;
 
 	do {
-		if (drv_data & FLAGS_VER2)
+		mdelay(100);
+		retry--;
+
+		if (drv_data & FLAGS_VER2) {
 			val = readl(&pdata->regs->regs_v2.tritsr);
-		else
+			/*
+			 * Check if TEMP is in valid range, the V bit in TRITSR
+			 * only reflects the RAW uncalibrated data
+			 */
+			valid =  ((val & 0xff) < 10 || (val & 0xff) > 125) ? 0 : 1;
+		} else {
 			val = readl(&pdata->regs->regs_v1.site[pdata->id].tritsr);
-	} while (!(val & 0x80000000));
+			valid = val & 0x80000000;
+		}
+	} while (!valid && retry > 0);
 
-	*temp = (val & 0xff) * 1000;
+	if (retry > 0) {
+		*temp = (val & 0xff) * 1000;
+	} else {
+		return -EINVAL;
+	}
 
 	return 0;
 }
-- 
2.38.0

